### PR TITLE
Handle CORS properly

### PR DESCRIPTION
**Pull request by jbrooksuk**
I'm using a Laravel package here to handle CORS properly. Could you confirm that this is working for you @joecohens?

### CODE REVIEW

File: config/cors.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+     | If you want them to apply to everything, you must define a path with *.
+     |
+     | allowedOrigins, allowedHeaders and allowedMethods can be set to ['*']
+     | to accept any value, the allowed methods however have to be explicitly listed.
+     |
+     */
+    'defaults' => [
+        'supportsCredentials' => true,
+        'allowedOrigins'      => [],
+        'allowedHeaders'      => [],
```
Comment (joecohens) [inline config/cors.php]: I guess we need to include `X-Cachet-Token` here to authenticate the API based on the package readme.

File: app/Providers/LoadConfigServiceProvider.php [hunk 2]
```diff
@@ -42,6 +42,25 @@ public function boot()
...
+            foreach ($domains as $domain) {
+                $allowedOrigins[] = $domain;
+            }
+        } else {
+            $allowedOrigins[] = getenv('APP_URL');
+        }
+
+        $this->app->config->set('cors.defaults.allowedOrigins', $allowedOrigins);
+
+        dd($this->app->config->get('cors.defaults'));
```
Comment (joecohens) [inline app/Providers/LoadConfigServiceProvider.php]: `dd` here lol

File: app/Providers/LoadConfigServiceProvider.php [hunk 3]
```diff
@@ -42,6 +42,25 @@ public function boot()
...
+        if ($allowedDomains = Setting::get('allowed_domains')) {
+            $domains = explode(',', $allowedDomains);
+            foreach ($domains as $domain) {
+                $allowedOrigins[] = $domain;
+            }
+        } else {
+            $allowedOrigins[] = getenv('APP_URL');
+        }
+
+        $this->app->config->set('cors.defaults.allowedOrigins', $allowedOrigins);
```
Comment (joecohens) [inline app/Providers/LoadConfigServiceProvider.php]: What do you think if we read the config first and then merge it with the DB values? So you can explicitly set them on the config file or on the DB.

File: app/Providers/LoadConfigServiceProvider.php [hunk 4]
```diff
@@ -42,6 +42,25 @@ public function boot()
         $this->app->config->set('app.url', $appDomain ?: $this->app->config->get('app.url'));
         $this->app->config->set('app.locale', $appLocale ?: $this->app->config->get('app.locale'));
 
+        // Setup Cors.
+        $allowedOrigins = [
+            Setting::get('app_domain'),
```
Comment (joecohens) [inline app/Providers/LoadConfigServiceProvider.php]: Remember to call Settings on the try catch if not we will have a failing test suite.

File: app/Providers/LoadConfigServiceProvider.php (hunk 3 above)
Comment (jbrooksuk) [inline app/Providers/LoadConfigServiceProvider.php]: Good idea! :)

**Comment by jbrooksuk** [comment]
Done @joecohens

File: config/cors.php (hunk 1 above)
Comment (jbrooksuk) [inline config/cors.php]: How do you mean?

File: app/Providers/LoadConfigServiceProvider.php (hunk 2 above)
Comment (jbrooksuk) [inline app/Providers/LoadConfigServiceProvider.php]: :laughing: whoops.

File: config/cors.php (hunk 1 above)
Comment (joecohens) [inline config/cors.php]: https://github.com/barryvdh/laravel-cors#configuration
> Note: When using custom headers, like X-Auth-Token or X-Requested-With, you must set the allowedHeaders to include those headers. You can also set it to array('*') to allow all custom headers.

Comment (jbrooksuk) [inline config/cors.php]: Gotcha. Will rebase with this on.

**Comment by jbrooksuk** [comment]
All done buddy.

**Comment by jbrooksuk** [comment]
@joecohens this looks to work now!

```
$ curl -I -H "Origin: https://api.cachet.com" --verbose http://cachet.io.dev/api/v1/incidents
```

@barryvdh thanks, looks like I wasn't testing with CORS via cURL properly.

**Comment by joecohens** [comment]
Awesome! :)

**Pull request merged**